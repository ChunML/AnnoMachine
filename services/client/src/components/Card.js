import React from 'react';

class Card extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      detectImage: true
    };
  }

  render() {
    const { image } = this.props;
    console.log(image)
    return (
      <div className="column">
        <div className="ui card" style={{margin: 'auto'}}>
          <div className="image" onClick={() => this.setState(prevState => ({detectImage: !prevState.detectImage}))}>
            <img src={`http://localhost:5000/${this.state.detectImage ? 'detects': 'uploads'}/${image.name}`} alt="Some image here" />
          </div>
          <div className="content">
            <div className="header">Author: { image.user.username }</div>
            <div className="meta">Uploaded at: { image.uploaded_at }</div>
            <div className="description">May contain: {image.boxes.length > 0 ?
              [...new Set(image.boxes.map(box => box.label))].map((label, id) => (
                <div key={ id } className='ui label' style={{background: '#11ee66'}}>{ label }</div>
              )) :
              <div className='ui label'>nothing</div>}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default Card;