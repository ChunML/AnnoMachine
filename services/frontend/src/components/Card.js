import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { FaClone, FaInfoCircle, FaTrashAlt } from 'react-icons/fa';

function Card({ image, currentUser, onDeleteImage }) {
  const [detectImage, toggleDetectImage] = useState(true);

  return (
    <div className="grid-item" style={{ textAlign: 'left' }}>
      <div className="card">
        <Link
          to={`/images/${image.name.replace('.jpg', '')}`}
          className="image"
        >
          <img
            src={`${process.env.REACT_APP_API_URL}/api/${
              detectImage ? 'detects' : 'uploads'
            }/${image.name}`}
            alt="detect result or original"
          />
        </Link>
        <div className="content">
          <div>
            <FaClone
              color={detectImage ? '#33ff33' : ''}
              onClick={() => toggleDetectImage(!detectImage)}
            />
            <Link
              to={`/images/${image.name.replace('.jpg', '')}`}
              className="ui image"
            >
              <FaInfoCircle color="#33ff33" />
            </Link>
            {currentUser === image.user.username && (
              <FaTrashAlt
                color="red"
                onClick={() => onDeleteImage(image.name.replace('.jpg', ''))}
              />
            )}
          </div>
          <div className="description">
            This image may contain:{' '}
            {image.boxes.length > 0 ? (
              [...new Set(image.boxes.map(box => box.label))].map(
                (label, id) => (
                  <span key={id} className="label">
                    {label}
                  </span>
                )
              )
            ) : (
              <div className="label">nothing</div>
            )}
          </div>
        </div>
        <div className="extra-content">
          <div>
            Uploaded by{' '}
            <span style={{ color: 'red' }}>{image.user.username}</span> at{' '}
            {image.uploaded_at}
          </div>
        </div>
      </div>
    </div>
  );
}

Card.propTypes = {
  image: PropTypes.object.isRequired,
  onDeleteImage: PropTypes.func.isRequired,
  currentUser: PropTypes.string.isRequired,
};

export default Card;
