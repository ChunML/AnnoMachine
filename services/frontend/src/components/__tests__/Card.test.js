import React from 'react';
import { MemoryRouter as Router } from 'react-router-dom';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Card from '../Card';
import '../../setupTests';

const image = {
  name: 'test.jpg',
  user: {
    username: 'testUser',
  },
  uploaded_at: '2000/01/01 01:02',
  boxes: [
    {
      label: 'dog',
    },
    {
      label: 'cat',
    },
    {
      label: 'dog',
    },
  ],
};

const imageNoBox = {
  name: 'test.jpg',
  user: {
    username: 'testUser',
  },
  uploaded_at: '2000/01/01 01:02',
  boxes: [],
};

const currentUser = ['testUser', 'testUser_2'];

describe('current user is the author of the image', () => {
  it('Card renders properly', () => {
    const onDeleteImage = jest.fn();
    const wrapper = shallow(
      <Card
        image={image}
        onDeleteImage={onDeleteImage}
        currentUser={currentUser[0]}
      />
    );
    const card = wrapper.find('.grid-item').find('.card');
    expect(card.length).toBe(1);
    const img = card.find('img');
    expect(img.length).toBe(1);
    expect(img.get(0).props.src).toContain(image.name);
    const faIcon = card.find('FaClone');
    expect(faIcon.length).toBe(1);
    faIcon.simulate('click');
    expect(wrapper.find('img').get(0).props.src).toContain('uploads');
    const faInfoCircle = card.find('FaInfoCircle');
    expect(faInfoCircle.length).toBe(1);
    const faTrashAlt = card.find('FaTrashAlt');
    expect(onDeleteImage).toHaveBeenCalledTimes(0);
    faTrashAlt.simulate('click');
    expect(onDeleteImage).toHaveBeenCalledTimes(1);
    const description = card.find('.description');
    expect(description.text()).toEqual('This image may contain: dogcat');
    const extraContent = card.find('.extra-content');
    expect(extraContent.text()).toBe(
      `Uploaded by ${image.user.username} at ${image.uploaded_at}`
    );
  });

  it('Card renders a snapshot properly', () => {
    const tree = renderer
      .create(
        <Router location="/">
          <Card
            image={image}
            onDeleteImage={jest.fn()}
            currentUser={currentUser[0]}
          />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('current user is not the author of the image', () => {
  it('Card renders properly', () => {
    const wrapper = shallow(
      <Card
        image={image}
        onDeleteImage={jest.fn()}
        currentUser={currentUser[1]}
      />
    );
    const card = wrapper.find('.grid-item').find('.card');
    expect(card.length).toBe(1);
    const img = card.find('img');
    expect(img.length).toBe(1);
    expect(img.get(0).props.src).toContain(image.name);
    const faIcon = card.find('FaClone');
    expect(faIcon.length).toBe(1);
    faIcon.simulate('click');
    expect(wrapper.find('img').get(0).props.src).toContain('uploads');
    const faInfoCircle = card.find('FaInfoCircle');
    expect(faInfoCircle.length).toBe(1);
    expect(wrapper.find('img').get(0).props.src).toContain('uploads');
    const description = card.find('.description');
    expect(description.text()).toEqual('This image may contain: dogcat');
    const extraContent = card.find('.extra-content');
    expect(extraContent.text()).toBe(
      `Uploaded by ${image.user.username} at ${image.uploaded_at}`
    );
  });

  it('Card renders a snapshot properly', () => {
    const tree = renderer
      .create(
        <Router location="/">
          <Card
            image={image}
            onDeleteImage={jest.fn()}
            currentUser={currentUser[1]}
          />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Image has no boxes', () => {
  it('Card renders properly', () => {
    const onDeleteImage = jest.fn();
    const wrapper = shallow(
      <Card
        image={imageNoBox}
        onDeleteImage={onDeleteImage}
        currentUser={currentUser[0]}
      />
    );
    const card = wrapper.find('.grid-item').find('.card');
    expect(card.length).toBe(1);
    const img = card.find('img');
    expect(img.length).toBe(1);
    expect(img.get(0).props.src).toContain(image.name);
    const faIcon = card.find('FaClone');
    expect(faIcon.length).toBe(1);
    faIcon.simulate('click');
    expect(wrapper.find('img').get(0).props.src).toContain('uploads');
    const faInfoCircle = card.find('FaInfoCircle');
    expect(faInfoCircle.length).toBe(1);
    const faTrashAlt = card.find('FaTrashAlt');
    expect(onDeleteImage).toHaveBeenCalledTimes(0);
    faTrashAlt.simulate('click');
    expect(onDeleteImage).toHaveBeenCalledTimes(1);
    const description = card.find('.description');
    expect(description.text()).toEqual('This image may contain: nothing');
    const extraContent = card.find('.extra-content');
    expect(extraContent.text()).toBe(
      `Uploaded by ${image.user.username} at ${image.uploaded_at}`
    );
  });

  it('Card renders a snapshot properly', () => {
    const tree = renderer
      .create(
        <Router location="/">
          <Card
            image={image}
            onDeleteImage={jest.fn()}
            currentUser={currentUser[0]}
          />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
