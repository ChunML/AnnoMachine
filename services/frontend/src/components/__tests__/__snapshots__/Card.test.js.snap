// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`current user is not the author of the image Card renders a snapshot properly 1`] = `
<div
  className="column"
  style={
    Object {
      "textAlign": "left",
    }
  }
>
  <div
    className="ui card"
    style={
      Object {
        "margin": "auto",
      }
    }
  >
    <a
      className="ui image"
      href="/images/test"
      onClick={[Function]}
    >
      <img
        alt="Some image here"
        src="undefined/api/detects/test.jpg"
      />
    </a>
    <div
      className="content"
    >
      <div
        style={
          Object {
            "marginBottom": "5px",
          }
        }
      >
        <i
          className="clone icon"
          onClick={[Function]}
          style={
            Object {
              "color": "#33ff33",
            }
          }
        />
        <a
          className="ui image"
          href="/images/test"
          onClick={[Function]}
        >
          <i
            className="hand point up outline icon"
            style={
              Object {
                "color": "#33ff33",
              }
            }
          />
        </a>
      </div>
      <div
        className="description"
      >
        This image may contain: 
        <div
          className="ui label"
          style={
            Object {
              "background": "#11ee66",
            }
          }
        >
          dog
        </div>
        <div
          className="ui label"
          style={
            Object {
              "background": "#11ee66",
            }
          }
        >
          cat
        </div>
      </div>
    </div>
    <div
      className="extra content"
    >
      <div>
        Uploaded by 
        <span
          style={
            Object {
              "color": "red",
            }
          }
        >
          testUser
        </span>
         at 
        2000/01/01 01:02
      </div>
    </div>
  </div>
</div>
`;

exports[`current user is the author of the image Card renders a snapshot properly 1`] = `
<div
  className="column"
  style={
    Object {
      "textAlign": "left",
    }
  }
>
  <div
    className="ui card"
    style={
      Object {
        "margin": "auto",
      }
    }
  >
    <a
      className="ui image"
      href="/images/test"
      onClick={[Function]}
    >
      <img
        alt="Some image here"
        src="undefined/api/detects/test.jpg"
      />
    </a>
    <div
      className="content"
    >
      <div
        style={
          Object {
            "marginBottom": "5px",
          }
        }
      >
        <i
          className="clone icon"
          onClick={[Function]}
          style={
            Object {
              "color": "#33ff33",
            }
          }
        />
        <a
          className="ui image"
          href="/images/test"
          onClick={[Function]}
        >
          <i
            className="hand point up outline icon"
            style={
              Object {
                "color": "#33ff33",
              }
            }
          />
        </a>
        <i
          className="trash alternate outline icon"
          onClick={[Function]}
          style={
            Object {
              "color": "red",
            }
          }
        />
      </div>
      <div
        className="description"
      >
        This image may contain: 
        <div
          className="ui label"
          style={
            Object {
              "background": "#11ee66",
            }
          }
        >
          dog
        </div>
        <div
          className="ui label"
          style={
            Object {
              "background": "#11ee66",
            }
          }
        >
          cat
        </div>
      </div>
    </div>
    <div
      className="extra content"
    >
      <div>
        Uploaded by 
        <span
          style={
            Object {
              "color": "red",
            }
          }
        >
          testUser
        </span>
         at 
        2000/01/01 01:02
      </div>
    </div>
  </div>
</div>
`;
